MSTR-50 : 문제 리스트 페이지의 필터에 태그 리스트가 보여진다.

Kim-Hyunjo opened the pull request:
![ezgif com-gif-maker (2)](https://user-images.githubusercontent.com/63906230/177522661-4890bb2e-78dd-4ce3-afa3-9bdc5431cee2.gif)

- 필터에서 선택한 태그 리스트가 보여짐

## 궁금한 점
- 문제가 250개 밖에 없다는 점에서 프론트단에서 태그에 따른 문제 필터링을 하는 것이 더 효율적일지? 아니면 백엔드에 요청을 보내서 리스트를 매번 받는게 나을지 고민입니다.

Kim-Hyunjo committed:
💄 Add li tag reset.css

- li tag 앞에 동그라미 없앰

Kim-Hyunjo committed:
♻️ Refactor TAGLIST

Kim-Hyunjo committed:
✨ Add Tag Component

Kim-Hyunjo committed:
🏷 Add Dropdown Props Type

Kim-Hyunjo committed:
♻️ Refactor checked tag list

- 선택된 태그 관리 question list page에서 하도록 리팩토링

Kim-Hyunjo committed:
✨ Add checked taglist in filter

- 선택된 태그 리스트 필터 하단에 보이도록 설정
- 선택된 태그리스트 전역으로 관리

Kim-Hyunjo committed:
🤖 Update TAGLIST constant

- 문제 풀이 kit 대신에 채점 가능 여부로 필터링하기로 했기 때문에 변경함

Kim-Hyunjo committed:
🔇 Remove log

kshired commented on src/Component/Utils/Dropdown/index.tsx:
```diff
@@ -12,38 +12,33 @@ interface ITagType {
       <div css={dropDownContentStyle(isOpen)}>
         <ul>
-          {tagType.type.map((e) => (
+          {type.map((e) => (
```
type이라는 키워드가 이미 typescript의 예약어로 알고 있는데, 변경하면 좋을 것 같아요

Kim-Hyunjo replied:
한 드롭다운의 element list를 나타내는 단어라 `elements` 로 변경했습니다!

kshired commented on src/Page/QuestionListPage.tsx:
```diff
@@ -21,10 +26,20 @@ function QuestionListPage() {
               ))}
             </div>
+            <ul css={checkedTagListStyle}>
+              {[...checkedTags].map((tagName) => (
```
질문 : `checkdTags`는  spread 연산자를 사용해야만 map을 쓸 수 있나요?

Kim-Hyunjo replied:
checkedTags는 중복을 방지하기 위해 Set 자료구조를 사용하였으므로 spread 연산자를 통해 Array 자료구조로 변환해야 Array.prototype.map()을 사용할 수 있습니다. 
항목 반복을 위해 forEach를 사용할 수는 있지만 return type이 void이므로 리액트 컴포넌트를 반환하지 않아 위 상황에서는 사용할 수 없습니다.

kshired replied:
@Kim-Hyunjo 

엥? 아래 코드 node에서 돌아가던데? 제가 잘못 알았나요..?

```js
const a = new Set([1,2,3]);
const b = a.map((i) => i+1);
console.log(b);
```

Kim-Hyunjo replied:
![스크린샷 2022-07-06 오후 7 58 58](https://user-images.githubusercontent.com/63906230/177535504-9c1f52df-a9dc-49a3-acf3-1fb4f536eae6.png)

앗 실행이 안되는데 혹시 다른 방법이 있을까요????

kshired replied:
@Kim-Hyunjo 제 착각인듯 ㅎㅎ 굿이에요!

kshired commented on src/constants/tag.ts:
```diff
@@ -20,37 +20,33 @@ const TAGLIST = {
       {
-        id: 'ds-1',
-        name: '자료구조 완벽정리 1',
+        id: 'grade-able',
```
gradable, non-gradable 과 같이 이미 사용되는 단어로 교체해도 좋을 듯 합니다.

Kim-Hyunjo replied:
안그래도 gradable이라는 단어를 찾아봤었는데 보통 형용사나 부사가 범위, 정도를 가지면 gradable하다라고 사용하는 경우가 많길래 애매해서 저렇게 넣었는데 다시 찾아보니 collins dictionary에는 "capable of being graded"라는 뜻이 있네요. 수정하겠습니다!!!

kshired commented on src/hooks/useStore.ts:
```diff
@@ -0,0 +1,22 @@
+  handleCheckedTags: (name: string, isChecked: boolean) => void;
+}
+
+// interface IStore extends ICheckedTags{
```
필요하지 않은 주석이면 지워도 될 듯 합니다.

kshired commented on src/constants/tag.ts:
```diff
@@ -20,37 +20,33 @@ const TAGLIST = {
-};
+];
 
 const TAGTYPELIST = ['category', 'isSolved', 'problemKit'] as const;
```
`TAGLIST`와의 일관성을 위해서 `,`를 마지막에 붙여보는건 어떨까요

Kim-Hyunjo replied:
오 이 부분을 수정하려다가 알았는데 원래 TAGTYPELIST를 키로 TAGLIST에서 가져와서 필터 부분을 렌더링하려고 했었는데 의미없이 복잡해져서 TAGLIST를 array로 바꾸고 TAGTYPELIST를 삭제했네요. 감사합니다 슨생님

kshired commented on src/Component/Tag/index.tsx:
```diff
@@ -0,0 +1,33 @@
+
+import { css } from '@emotion/react';
+
+interface TTag {
```
다른 interface는 `I${타입이름}` 인 것 같은데, 얘는 왜 TT일까요

Kim-Hyunjo replied:
??? 그러게요???? 역시 1px의 마스터...!!!!!!

Kim-Hyunjo committed:
🚚 Rename type -> elements

- Dropdown 컴포넌트의 prop인 type이 ts 예약어임
- Dropdown내의 리스트 요소들이라는 의미를 살려 elements로 변경

Kim-Hyunjo committed:
🔥 Remove Unused codes

Kim-Hyunjo committed:
✏️ Fix typo

- interface convention에 맞게 I로 시작하도록 네이밍 변경

Kim-Hyunjo committed:
💄 Update QuestionListPage UI

- 태그 크기 키움
- 태그 리스트 너비 설정

kshired commented on src/Page/QuestionListPage.tsx:
```diff
@@ -120,10 +120,11 @@ const dropdownListStyle = css`
   display: grid;
-  grid-template-columns: repeat(auto-fill, minmax(20vw, 1fr));
-  grid-gap: 1rem;
+  grid-template-columns: repeat(auto-fill, minmax(60px, 1fr));
```
얘네는 rem, vw이였다가 왜 px로 바뀌었어요?

Kim-Hyunjo replied:
grid layout로 반응형을 적용해보려고 추가해봤는데 어림도 없게 안돼서 문제 리스트 페이지 반응형 레이아웃은 추후에 구현할 예정입니다...
따라서 원래대로 돌려놨읍니다..
공부해보고 적용하겠습니다

kshired commented on src/Component/Tag/index.tsx:
```diff
@@ -2,11 +2,11 @@
 }
 
-function Tag({ name }: TTag) {
+function Tag({ name }: ITag) {
```
생각해보니까 `name`만 parameter로 받는 함수인데 굳이 interface를 쓸 필요가 있나요?
추후 데이터가 늘어날 수도 있어서 일단 선언해놓은 걸까요?

Kim-Hyunjo replied:
이유1. Tag component의 prop이 name이라는 사실을 ts가 받아들이지 못함
`(alias) function Tag(name: string): EmotionJSX.Element
import Tag
Type '{ name: string; key: string; }' is not assignable to type 'string'`

이유2. 나중에 컬러도 속성에 추가할것입니다.

감사합니다 슨샌님

kshired reviewed (approved):
![UntidyDistinctAiredaleterrier-max-1mb](https://user-images.githubusercontent.com/36851531/177671851-894ac278-d690-4e1d-869f-a70ec7205c07.gif)
